python-package/xgboost/collective.py

```diff
@@ -0,0 +1,244 @@
+    is_dist = _LIB.XGCommunicatorIsDistributed()
+    return is_dist
+
+
+def communicator_print(msg: Any) -> None:
+    """Print message to the communicator.
+
+    This function can be used to communicate the information of
+    the progress to the communicator.
+
+    Parameters
+    ----------
+    msg : str
+        The message to be printed to the communicator.
+    """
+    if not isinstance(msg, str):
+        msg = str(msg)
+    is_dist = _LIB.XGCommunicatorIsDistributed()
+    if is_dist != 0:
+        _check_call(_LIB.XGCommunicatorPrint(c_str(msg)))
+    else:
+        print(msg.strip(), flush=True)
+
+
+def get_processor_name() -> str:
+    """Get the processor name.
+
+    Returns
+    -------
+    name : str
+        the name of processor(host)
+    """
+    name_str = ctypes.c_char_p()
+    _check_call(_LIB.XGCommunicatorGetProcessorName(ctypes.byref(name_str)))
+    value = name_str.value
+    assert value
+    return py_str(value)
+
+
+def broadcast(data: _T, root: int) -> _T:
+    """Broadcast object from one node to all other nodes.
+
+    Parameters
+    ----------
+    data : any type that can be pickled
+        Input data, if current rank does not equal root, this can be None
+    root : int
+        Rank of the node to broadcast data from.
+
+    Returns
+    -------
+    object : int
+        the result of broadcast.
+    """
+    rank = get_rank()
+    length = ctypes.c_ulong()
+    if root == rank:
+        assert data is not None, 'need to pass in data when broadcasting'
+        s = pickle.dumps(data, protocol=pickle.HIGHEST_PROTOCOL)
+        length.value = len(s)
+    # run first broadcast
+    _check_call(_LIB.XGCommunicatorBroadcast(ctypes.byref(length),
+                                             ctypes.sizeof(ctypes.c_ulong), root))
+    if root != rank:
+        dptr = (ctypes.c_char * length.value)()
+        # run second
+        _check_call(_LIB.XGCommunicatorBroadcast(ctypes.cast(dptr, ctypes.c_void_p),
+                                                 length.value, root))
+        data = pickle.loads(dptr.raw)
+        del dptr
+    else:
+        _check_call(_LIB.XGCommunicatorBroadcast(ctypes.cast(ctypes.c_char_p(s), ctypes.c_void_p),
+                                                 length.value, root))
+        del s
+    return data
+
+
+# enumeration of dtypes
+DTYPE_ENUM__ = {
+    np.dtype('int8'): 0,
+    np.dtype('uint8'): 1,
+    np.dtype('int32'): 2,
+    np.dtype('uint32'): 3,
+    np.dtype('int64'): 4,
+    np.dtype('uint64'): 5,
+    np.dtype('float32'): 6,
+    np.dtype('float64'): 7
+}
+
+
+@unique
+class Op(IntEnum):
+    """Supported operations for rabit."""
+    MAX = 0
+    MIN = 1
+    SUM = 2
+
+
+def allreduce(  # pylint:disable=invalid-name
+        data: np.ndarray, op: Op
+) -> np.ndarray:
+    """Perform allreduce, return the result.
+
+    Parameters
+    ----------
+    data :
+        Input data.
+    op :
+        Reduction operators, can be MAX or SUM
```

Comment:
Not `MIN`? I think we can omit `can be ...` in the doc since we have the Python type hint.

Reply:
We don't currently use `MIN`, but added it back since it's not too much code.